openff/toolkit/topology/molecule.py

```diff
@@ -6629,3 +6722,149 @@ def _ipython_display_(self):
+                else:
+                    particle_key.append("None")
+
+            hier_eles_to_add[tuple(particle_key)].append(particle)
+
+        # Create the actual HierarchyElements
+        for particle_key, particles_to_add in hier_eles_to_add.items():
+            particle_indices = [p.molecule_particle_index for p in particles_to_add]
+            self.add_hierarchy_element(particle_key, particle_indices)
+
+        self.sort_hierarchy_elements()
+
+    def add_hierarchy_element(self, identifier, particle_indices):
+        """
+        Instantiate a new HierarchyElement belonging to this HierarchyScheme.
+        This is the main way to instantiate new HierarchyElements.
+
+        Parameters
+        ----------
+        identifier : tuple of str and int
+            uniqueness tuple
+        particle_indicess : iterable int
+        """
+        new_hier_ele = HierarchyElement(self, identifier, particle_indices)
+        self.hierarchy_elements.append(new_hier_ele)
+        return new_hier_ele
+
+    def sort_hierarchy_elements(self):
+        """
+        Semantically sort the HierarchyElements belonging to this object, according to
+        their identifiers.
+        """
+        # hard-code the sort_func value here, since it's hard to serialize safely
+        sort_func = lambda x: version.parse(".".join([str(i) for i in x.identifier]))
+        self.hierarchy_elements.sort(key=sort_func)
+
+
+class HierarchyElement:
+    def __init__(self, scheme, identifier, particle_indices):
+        """
+        scheme : HierarchyScheme
+        id : tuple of str and int
+            uniqueness tuple
+        particle_indicess : iterable int
+        """
+        self.scheme = scheme
+        self.identifier = identifier
+        self.particle_indices = deepcopy(particle_indices)
+        for id_component, uniqueness_component in zip(
+            identifier, scheme.uniqueness_criteria
+        ):
+            setattr(self, uniqueness_component, id_component)
+
+    def to_dict(self):
+        """
+        Serialize this object to a basic dict of strings, ints, and floats
+        """
+        return_dict = dict()
+        return_dict["identifier"] = self.identifier
+        return_dict["particle_indices"] = self.particle_indices
+        return return_dict
+
+    @property
+    def particles(self):
+        for particle_index in self.particle_indices:
+            yield self.particles[particle_index]
+
+    def particle(self, index: int):
+        """
+        Get particle with a specified index.
+
+        Parameters
+        ----------
+        index : int
+
+        Returns
+        -------
+        particle : openff.toolkit.topology.Particle
+        """
+        return self._particles[index]
+
+    @property
+    def parent(self):
+        return self.scheme.parent
+
+    def __str__(self):
+        return f"HierarchyElement {self.identifier} containing {len(self.particles)} particles"
```

Comment:
Just a detail here, `self.particles` is a generator so applying `len` to it returns a `TypeError`. Now, I wonder if we should add tests for these things (maybe some tests that are specific to `HierarchyElement` and `HierarchyScheme` classes/objects), what do you think?

Reply:
Wow!  Thanks for the close review :-) Fixing that uncovered a ton of other bugs, and now I've fixed those and added more tests. 

Unfortunately, the way it's designed, we can't test `HierarchySchemes` or `HierarchyElements` individually, since we can't instantiate them without a parent `Molecule` (for schemes) or `HierarchyScheme` (for elements). 

(The OpenFF `Atom` class is a good example of what happens if we allow users to treat parents as optional. An `Atom` CAN be instantiated without a `Molecule` (it's allowed to be `None`), but about half of API points just throw errors if a user tries to access them, since they need information from the `self.molecule`. So that's a design decision I regret from earlier in OpenFF)